### DISCUSSION TITLE
停留点删除

### DISCUSSION THREAD

**Discussion by Murphy666-code** [zdsjjtTLG/TrackIt#21] [Q&A]
进行轨迹清洗 想处理一下距离比较近的重复点 使用停留点删除设置 tp.del_dwell_points(dwell_l_length=10, dwell_n=2)但是出来之后点位数量没有减少、没有达到预期要求

**Comment by Murphy666-code** [comment]
![微信图片_20241230152142](https://github.com/user-attachments/assets/217c443a-664c-4dab-a728-104a5cdff6d2)

**Reply by zdsjjtTLG** [reply]
那你可以试试调整参数呀，比如把dwell_l_length调高, dwell_n调低，这样不就更加容易满足删除条件嘛？

**Reply by Murphy666-code** [reply]
参数已经根据我的gps数据调整过了 比如图右上方三个挨着的点距离不超过五米 所以我设置的是十米 N=2 但最后出来的结果是还是三个点，只是位置稍微调整了下

**Reply by zdsjjtTLG** [reply]
dwell_n调到0呢？

**Reply by Murphy666-code** [reply]
啊 还可以是0啊 我以为最小得是2 我看之前英文版的解释是dwell_n：如果连续dwell_n个以上的相邻 GPS 点之间的距离小于 dwell_l_length，则这组点将被标识为一个停留点，默认值为 2

**Reply by zdsjjtTLG** [reply]
应该是连续n+1个点，文档应该写错了，你再试试n=0或者n=1

**Reply by Murphy666-code** [reply]
哦哦 可以了 感谢

**Reply by zdsjjtTLG** [reply]
别看英文版，N久没维护了，看中文的把

**Reply by Murphy666-code** [reply]
现在还是以看中文为主 但中文的没找到相关参数的解释 所以去翻的英文

**Reply by zdsjjtTLG** [reply]
你点击一下这个函数他就链接到函数接口了，参数解释都在那里：
![image](https://github.com/user-attachments/assets/5f4b92e0-7a83-4750-88ee-361b1d28b361)

**Reply by Murphy666-code** [reply]
好哒 谢谢

**Reply by Murphy666-code** [reply]
轨迹点增密可以基于路网增密吗，我的gps数据点位之间跨度比较大 最远的能差五分钟 试了一下轨迹点增密发现是两点之间连线上均匀增加点位但是大部分不在路网上

**Reply by Murphy666-code** [reply]
因为我看最后的地图匹配代码里有停留点删除、增密等，这里的操作和地图匹配前轨迹点预处理中的停留点删除和增密等有区别吗
mpm = MapMatch(net=my_net, flag_name='xa_sample', # 指定项目名称xa_sample
               use_sub_net=True, # 启用子网络
               gps_buffer=800, top_k=80, # GPS点空间关联半径取100米，选取GPS点100米范围内最近的20个路段作为候选路段
               dense_gps=False, # 不增密GPS点
               use_heading_inf=True, omitted_l=6.0,  # 启用GPS航向角矫正，若前后点GPS距离<=6米，将忽略航向角的修正
               del_dwell=True, dwell_l_length=50.0, dwell_n=0,  # 停留点删除参数
               export_html=True, export_geo_res=True, use_gps_source=True,  # 输出设置参数
               gps_radius=15.0, export_all_agents=False, # 输出设置参数
               out_fldr=r'D:\地图匹配实验2')  # 输出设置参数

**Reply by zdsjjtTLG** [reply]
1.地图匹配里面自带的预处理(无滤波和简化)和轨迹清洗里面的预处理是一样的；
2.在没有进行路径匹配之前，我根本不知道他沿哪条路走的，怎么可能给你按照路径增密呢，你这种稀疏的，可以不用增密，把cut_off设的大一些就好了，他会把路径搜索出来的

**Reply by Murphy666-code** [reply]
回复错位置了，见下个评论

**Comment by Murphy666-code** [comment]
您好 元旦快乐！今天在尝试调参数 遇到了点问题
1、经过多次尝试发现buffer调整对报错数量的减少效果相对显著，相较于cut off来说，但带来的问题是gps点位匹配后偏移距离较大。
2、对报错部分路网进行检查并没有发现有明显问题，路网前期也经过了代码的连通性修复。
以下是我的运行代码：
import pandas as pd
import geopandas as gpd
from gotrackit.map.Net import Net
from gotrackit.MapMatch import MapMatch

link = gpd.read_file(r'D:\python处理\modifiedConnLink.shp')
node = gpd.read_file(r'D:\python处理\modifiedConnNode.shp')
my_net = Net(link_gdf=link, node_gdf=node, cut_off=5000.0)
my_net.init_net()  # net初始化
gps_df = pd.read_csv(r'D:\数据清洗实验.csv')
mpm = MapMatch(net=my_net, flag_name='xa_sample', # 指定项目名称xa_sample
               use_sub_net=True, # 启用子网络
               gps_buffer=500, top_k=200, # GPS点空间关联半径取100米，选取GPS点100米范围内最近的20个路段作为候选路段
               dense_gps=False, # 不增密GPS点
               use_heading_inf=False, omitted_l=6.0,  # 启用GPS航向角矫正，若前后点GPS距离<=6米，将忽略航向角的修正
               del_dwell=True, dwell_l_length=50.0, dwell_n=0,  # 停留点删除参数
               export_html=True, export_geo_res=True, use_gps_source=True,  # 输出设置参数
               gps_radius=15.0, export_all_agents=False, # 输出设置参数
               out_fldr=r'D:\地图匹配实验2')  # 输出设置参数

match_res, warn_info, error_info = mpm.execute(gps_df=gps_df)
match_res.to_csv(r'D:\地图匹配实验2\match_res.csv', encoding='utf_8_sig', index=False)
结果如下：
[地图匹配实验2.zip](https://github.com/user-attachments/files/18289214/2.zip)
以上是挑选的一辆车进行的实验

**Reply by Murphy666-code** [reply]
现在难点就是大幅度调buffer可以解决报错问题，但是匹配结果精度可能就受影响，小幅度调的话就会有大量报错以及有的点位之间连不上的问题。不知道还有没有好的解决方法

**Reply by zdsjjtTLG** [reply]
你的数据很稀疏且没有**切分行程**(你给的那个样例应该是多次出行，他们不应该作为一条出行链来考虑，你应该先切分行程)，另外你这个情况还是开一下增密吧，cut_off就不用那么大了，top_k也没必要调到500，你按照这个试试，看看情况如何。

**Reply by zdsjjtTLG** [reply]
其实你说的有些点没有匹配到最近的路段，那个是因为走远处的路段时转移概率更大，你可以尝试降低gps_sigma，来提升他对**发射概率中距离的敏感性**，当然，你还是先按照我刚刚说的做，那个是标准的流程

**Reply by Murphy666-code** [reply]
您好 我这个是按照手册操作过一遍了 设置了30min时间阈值进行行程切分 我这个是网约车轨迹数据 所以可能看起来像是没有切分行程

**Reply by Murphy666-code** [reply]
如果要调gps sigma的话就直接在代码里buffer后加上一行gps_sigma=数值就行是吧

**Reply by zdsjjtTLG** [reply]
所有的可调参数都在文档的函数接口里面写了解释

**Reply by zdsjjtTLG** [reply]
网约车上一单和下一单隔的很近，30分钟太大了吧
